### PR TITLE
Add cache miss stats/metrics to server

### PR DESCRIPTION
**Pull request by rmccorm4**
Server
- [x] Add to prometheus metrics
- [x] Add to infer stats
- [x] Add to HTTP and GRPC server stats responses
- [x] Update `docs/metrics.md`
- [x] Update `docs/protocol/extension_statistics.md`

Non-server
- [x] Add cache_miss to common GRPC protobuf stub: https://github.com/triton-inference-server/common/pull/42
- [x] Separate cache_miss stat out from "overhead" in perf_analyzer output when caching enabled: https://github.com/triton-inference-server/client/pull/70

### CODE REVIEW

File: src/core/dynamic_batch_scheduler.cc [hunk 1]
```diff
@@ -582,12 +582,24 @@ DynamicBatchScheduler::DelegateResponse(
   completion_queue_.emplace_back();
   auto queue_slot = &completion_queue_.back();
   uint64_t request_hash = request->CacheKey();
+  // Pass raw ptr to lambda for tracking stats from cache and updating
+  // metric reporter on cache miss stats after insertion
+  InferenceRequest* raw_request_ptr = request.get();
+
   request->SetResponseDelegator(
-      [this, queue_slot, request_hash](
+      [this, queue_slot, request_hash, raw_request_ptr](
```
Comment (rmccorm4) [inline src/core/dynamic_batch_scheduler.cc:590]: Not sure if there's a better way to do this. Passing the unique_ptr has an ownership/deletion issue

**Comment by rmccorm4** [comment]
Pipeline #4047871

File: src/core/response_cache.cc [hunk 2]
```diff
@@ -225,39 +154,104 @@ RequestResponseCache::Insert(
...
+    CaptureInsertionEndTime(request);
     return Status(
         Status::Code::ALREADY_EXISTS,
         "key [" + std::to_string(key) + "] already exists in cache");
   }
 
   // Construct cache entry from response
   auto entry = CacheEntry();
-  RETURN_IF_ERROR(BuildCacheEntry(response, &entry));
+  auto status = BuildCacheEntry(response, &entry);
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/core/response_cache.cc:210]: Added `ScopedTimer` class see below

File: src/core/response_cache.cc [hunk 3]
```diff
@@ -195,53 +167,135 @@ RequestResponseCache::Lookup(
   // Populate passed-in "ptr" from cache entry
   auto entry = iter->second;
   // Build InferenceResponse from CacheEntry
-  RETURN_IF_ERROR(BuildInferenceResponse(entry, ptr));
```
Comment (tanmayv25) [inline src/core/response_cache.cc:198]: Is this breakdown necessary?
I believe it only adds the following message before error:
"Failed to build inference response: "

File: src/core/response_cache.cc [hunk 4]
```diff
@@ -195,53 +167,135 @@ RequestResponseCache::Lookup(
...
   if (iter != cache_.end()) {
     return Status(
-        Status::Code::INTERNAL,
+        Status::Code::ALREADY_EXISTS,
         "key [" + std::to_string(key) + "] already exists in cache");
   }
 
   // Construct cache entry from response
   auto entry = CacheEntry();
-  RETURN_IF_ERROR(BuildCacheEntry(response, &entry));
```
Comment (tanmayv25) [inline src/core/response_cache.cc:230]: Same here. I think we can just leave it as :
RETURN_IF_ERROR(BuildCacheEntry(response, &entry));

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**